tutorials/tour/higher-order-functions.md

```diff
@@ -8,34 +8,112 @@ tutorial: scala-tour
+
+## Functions that accept functions
+One reason to use higher-order functions is to reduce redundant code. Let's say you wanted functions  that could either search for files by directory, by regular expression, or by a substring. Without creating a higher-order function,
+it might look something like this:
 
+```tut
+  object FileFinder {
+    private def filesHere = (new java.io.File(".")).listFiles
+
+    def filesEndingWith(query: String) =
+      for (file <- filesHere; if file.getName.endsWith(query))
+        yield
+
+    def filesContaining(query: String) =
+      for (file <- filesHere; if file.getName.contains(query))
+        yield file
+
+    def filesMatchingRegex(query: String) =
+      for (file <- filesHere; if file.getName.matches(query))
+        yield file
+  }
 ```
-[7]
+
+Notice how each of the three methods vary only by a single method call. To simplify,
+you can extract the repeated code into a higher-order function like so:
+
+```tut
+object FileMatcher {
+  private def filesHere = (new java.io.File(".")).listFiles
+
+  private def filesMatching(matcher: String => Boolean) =
+    for (file <- filesHere; if matcher(file.getName))
+          yield file
+
+    def filesEndingWith(query: String) =
+      filesMatching(_.endsWith(query))
+
+    def filesContaining(query: String) =
+      filesMatching(_.contains(query))
+
+    def filesMatchingRegex(query: String) =
+      filesMatching(_.matches(query))
+    }
+```
+
+The new function, `filesMatching`, takes a function of type `String => Boolean`
+(i.e. a function that takes a String and returns a Boolean) and returns the Sequence comprehension
+created by the `for`/`yield`.
+
+Credit: Odersky, Martin, Lex Spoon, and Bill Venners. Programming in Scala. Walnut Creek, CA: Artima, 2010. Web.
+
+## Functions that return functions
+
+There are certain cases where you want to generate a function. Here's an example
+of a method that returns a function.
+
+```tut
+def urlBuilder(ssl: Boolean, domainName: String): (String, String) => String = {
+  val schema = if (ssl) "https://" else "http://"
+  (endpoint: String, query: String) => s"$schema$domainName/$endpoint?$query"
+}
+​
+val domainName = "www.example.com"
+def getURL = urlBuilder(ssl=true, domainName)
+​
+val endpoint = "users"
+val query = "id=1"
+val url = getURL(endpoint, query) // "https://www.example.com/users?id=1": String
+​
 ```
 
-In this example, the method `decorator.layout` is coerced automatically to a value of type `Int => String` as required by method `apply`. Please note that method `decorator.layout` is a _polymorphic method_ (i.e. it abstracts over some of its signature types) and the Scala compiler has to instantiate its method type first appropriately.
+Notice the return type of urlBuilder `(String, String) => String`. This means that
+the returned anonymous function takes two Strings and returns a String.
```

Comment:
Maybe you should emphasize that the syntax at the type level is homogeneous with the syntax at the value level.